pkg/operator/csidriveroperator/driverstarter.go

```diff
@@ -0,0 +1,149 @@
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceapply"
+	"github.com/openshift/library-go/pkg/operator/staticresourcecontroller"
+	"github.com/openshift/library-go/pkg/operator/status"
+	"github.com/openshift/library-go/pkg/operator/v1helpers"
+)
+
+const (
+	infraConfigName = "cluster"
+)
+
+// This CSIDriverStarterController starts CSI driver controllers based on the
+// underlying cloud. It does not install anything by itself, only monitors
+// Infrastructure instance and starts individual ControllerManagers for the
+// particular cloud. It produces following Conditions:
+// CSIDriverStarterDegraded - error checking the Infrastructure
+type CSIDriverStarterController struct {
+	operatorClient v1helpers.OperatorClient
+	infraLister    openshiftv1.InfrastructureLister
+	versionGetter  status.VersionGetter
+	targetVersion  string
+	eventRecorder  events.Recorder
+
+	controllers []csiDriverControllerManager
+}
+
+type csiDriverControllerManager struct {
+	name           string
+	operatorConfig csioperatorclient.CSIOperatorConfig
+	platform       configv1.PlatformType
+	mgr            manager.ControllerManager
+	running        bool
+}
+
+func NewCSIDriverStarterController(
+	clients *csoclients.Clients,
+	resyncInterval time.Duration,
+	versionGetter status.VersionGetter,
+	targetVersion string,
+	eventRecorder events.Recorder,
+	driverConfigs []csioperatorclient.CSIOperatorConfig) factory.Controller {
+	c := &CSIDriverStarterController{
+		operatorClient: clients.OperatorClient,
+		infraLister:    clients.ConfigInformers.Config().V1().Infrastructures().Lister(),
+		versionGetter:  versionGetter,
+		targetVersion:  targetVersion,
+		eventRecorder:  eventRecorder,
+	}
+
+	// Populating all CSI driver operator ControllerManagers here simplifies
+	// the startup a lot
+	// - All necessary informers are populated.
+	// - CSO then just Run()s
+	//All ControllerManagers are nor running at this point! They will be started
+	c.populateCSIControllers(clients, resyncInterval, driverConfigs)
+
+	return factory.New().WithSync(c.sync).WithSyncDegradedOnError(clients.OperatorClient).WithInformers(
+		clients.OperatorClient.Informer(),
+		clients.ConfigInformers.Config().V1().Infrastructures().Informer(),
+	).ToController("CSIDriverStarter", eventRecorder)
+}
+
+func (c *CSIDriverStarterController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	opSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+	if opSpec.ManagementState != operatorapi.Managed {
+		return nil
+	}
+
+	infrastructure, err := c.infraLister.Get(infraConfigName)
+	if err != nil {
+		return err
+	}
+
+	platform := infrastructure.Status.PlatformStatus.Type
+	for i := range c.controllers {
+		if c.controllers[i].operatorConfig.Platform == platform && !c.controllers[i].running {
+			go c.controllers[i].mgr.Start(ctx)
+			c.controllers[i].running = true
```

Comment:
This isn't ever going to be false anymore, do we care?

When the `ctx` above is cancelled, all controllers from `c.controllers[i].mgr` will shutdown (which is what we want). However, `running` will still be true, and nothing sets it back to false. If for some reason we reach line 93 above again, the CSI driver operator will never run again.

I don't know if this can ever happen, but I wish we could get rid of `.running` and have the ControllerManager tell us if it's running or not (maybe we can).

Reply:
I don't think we really need to solve this. The context gets cancelled at the operator shutdown and we won't ever want to statt the CSI driver operators again.